src/nexmo/errors.py

```diff
@@ -12,3 +14,137 @@ class ServerError(Error):
 
 class AuthenticationError(ClientError):
     pass
+
+class NexmoError(ClientError):
+    pass
+
+#Code 1: Throttled
+class ThrottledError(NexmoError):
+    pass
+
+#Code 2: Missing params
+class MissingParamError(NexmoError):
+    pass
+
+#Code 3: Invalid params
+class InvalidParamError(NexmoError):
+    pass
+
+#Code 4: Invalid credentials
+class CredentialError(NexmoError):
+    pass
+
+#Code 5: Internal Error
+class InternalError(NexmoError):
+    pass
+
+#Code 6: Invalid message
+class InvalidMessageError(NexmoError):
+    pass
+
+#Code 7: Number barred
+class NumberBarredError(NexmoError):
+    pass
+
+#Code 8: Partner account barred
+class PartnerAccountBarredError(NexmoError):
+    pass
+
+#Code 9: Partner quota exceeded
+class PartnerQuotaExceededError(NexmoError):
+    pass
+
+#Code 11: Account not enabled for REST
+class AccountNoRestError(NexmoError):
+    pass
+
+#Code 12: Message too long
+class MessageLengthError(NexmoError):
+    pass
+
+#Code 13: Communication Failed
+class CommunicationError(NexmoError):
+    pass
+
+#Code 14: Invalid Signature
+class InvalidSignatureError(NexmoError):
+    pass
+
+#Code 15: Illegal Sender Address - rejected
+class IllegalSenderError(NexmoError):
+    pass
+
+#Code 16: Invalid TTL
+class InvalidTTLError(NexmoError):
+    pass
+
+#Code 19: Facility not allowed
+class FacilityError(NexmoError):
+    pass
+
+#Code 20: Invalid Message class
+class InvalidMessageClassError(NexmoError):
+    pass
+
+#Code 23: Bad callback :: Missing Protocol
+class MissingProtocolError(NexmoError):
+    pass
+
+#Code 29: Non White-listed Destination
+class BlackListDestinationError(NexmoError):
+    pass
+
+#Code 34: Invalid or Missing Msisdn Param
+class MsisdnError(NexmoError):
+    pass
+
+#Code 400: Bad request - when voice Api
+class BadRequestError(NexmoError):
+    def __init__(self, data):
+        error_data = data
+        message = data
+        if isinstance(data, bytes):
+            #convert data bytes in dict and re-assign
+            error_data = json.loads(data.decode('utf-8'))
+        if "type" in error_data and "title" in error_data:
+            if "detail" in error_data:
+                message = "{title}: {detail} ({type})".format(
+                        title=error_data["title"], detail=error_data["detail"], type=error_data["type"]
+                )
+            elif "invalid_parameters" in error_data:
+                message = "{title}: Invalid parameters, reason: {reason}, param hint: {name}".format(
+                    title=error_data["title"], reason=error_data["invalid_parameters"][0]["reason"], name=error_data["invalid_parameters"][0]["name"]
+                )
+        super().__init__(message)
+
+class ExceptionHandler():
```

Comment:
Should this really be a generic "ExceptionHandler" - this looks like it's pretty specifically an SMS exception handler - which has its own virtue but should probably not be named so generically - again this goes back to how individual errors should be handled.

Reply:
@slorello89 It has a generic name because it's used in parse which is a generic method for both sms and voice. Depending on the error code a specific exception is generated that's the reason behind naming it that way. Exception handler being a handler that has a collection of exceptions and each triggers the specific exception depending on the error code regardless where it comes from.